framework/include/controls/ControlInterface.h

```diff
@@ -0,0 +1,219 @@
+/****************************************************************/
+
+#ifndef CONTROLINTERFACE_H
+#define CONTROLINTERFACE_H
+
+// MOOSE includes
+#include "InputParameters.h"
+#include "InputParameterWarehouse.h"
+#include "MooseUtils.h"
+
+// Forward declarations
+class ControlInterface;
+
+template<>
+InputParameters validParams<ControlInterface>();
+
+/**
+ * An interface class for accessing writable references to InputParameters
+ *
+ * This class exists to provided parameter access to the Control base class
+ * as well as the RealParameterReporter postprocessor, no other objects
+ * should use this class. Hence, the private constructor
+ */
+class ControlInterface
+{
+protected:
+
+  /**
+   * Obtain a writeable reference to a single parameter given a InputParameter name
+   */
+  template<typename T>
+  T & getControlParam(const std::string & name);
+
+  /**
+   * Obtain a writeable reference to a single parameter given a specific name
+   */
+  template<typename T>
+  T & getControlParamByName(const std::string & name);
+
+  /**
+   * Obtain pointers to all parameters given a InputParameter name
+   */
+  template<typename T>
+  std::vector<T*> getControlParamVector(const std::string & name);
+
+  /**
+   * Obtain pointers to all parameters given a specific name
+   */
+  template<typename T>
+  std::vector<T*> getControlParamVectorByName(const std::string & name);
+
+  /**
+   * Destructor
+   */
+  virtual ~ControlInterface(){}
+
+private:
+
+  /**
+   * A private constructor to limit the use of this interface to Control related objects
+   */
+  ControlInterface(const InputParameters & parameters);
+
+  /**
+   * Helper method for extracting iterators to InputParameter objects containing a property with the given name
+   * @param name The complete name of the desired parameter (only used for error messages)
+   * @param desired The desired parameter separated into components via tokenizeName method of this class
+   *
+   * Using the iterator allows for the public get methods to provide better error messages because the
+   * full object name can be reported.
+   *
+   * @see tokenizeName
+   */
+  template<typename T>
+  std::vector<InputParameterIterator> getValidInputParameterIteratorsByName(const std::string & name, const std::deque<std::string> & desired);
+
+  /**
+   * Helper method separating parameter name into system/group/param format
+   *
+   * [System]
+   *   [./group]
+   *     param = ...
+   *
+   */
+  std::deque<std::string> tokenizeName(const std::string & name);
+
+  /// Reference to the Control objects parameters
+  const InputParameters & _ci_parameters;
+
+  /// A Reference to MooseApp (needed to gain access to InputParameterWarehouse)
+  MooseApp & _ci_app;
+
+  /// A reference to the InputParameterWarehouse which is used for access the parameter objects
+  InputParameterWarehouse & _input_parameter_warehouse;
+
+  /// The object thread id, needed for access the warehouse
+  THREAD_ID _tid;
```

Comment:
I agree with @friedmud on this, I just don't think I can get all this
cleaned up in a week and add the python calling stuff raven needs for the
milestone.

On Thu, Sep 10, 2015 at 9:31 AM, Derek Gaston <notifications@github.com>
wrote:

> In framework/include/controls/ControlInterface.h
> <https://github.com/idaholab/moose/pull/5676#discussion_r39174155>:
>
> > +   *     param = ...
> > +   *
> > +   */
> > +  std::deque<std::string> tokenizeName(const std::string & name);
> > +
> > +  /// Reference to the Control objects parameters
> > +  const InputParameters & _ci_parameters;
> > +
> > +  /// A Reference to MooseApp (needed to gain access to InputParameterWarehouse)
> > +  MooseApp & _ci_app;
> > +
> > +  /// A reference to the InputParameterWarehouse which is used for access the parameter objects
> > +  InputParameterWarehouse & _input_parameter_warehouse;
> > +
> > +  /// The object thread id, needed for access the warehouse
> > +  THREAD_ID _tid;
>
> CL shouldn't be threaded.
>
> How to deal with threads with CL is another topic.
>
> Right now we store a copy of InputParameters for each thread for each
> object. We've talked a lot about doing away with that... and now that
> InputParameters return const references... it may be time to do just that.
> What this would mean is that the InputParameter storage would only store
> *one* copy of InputParameters that would be shared by all threaded copies
> of an object. This would make CL a billion times simpler as it only needs
> to update the one value...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5676/files#r39174155>.
>